Source/Core/Core/CoreTiming.cpp

```diff
@@ -326,6 +341,76 @@ void CoreTimingManager::Advance()
   PowerPC::CheckExternalExceptions();
 }
 
+void CoreTimingManager::Throttle(const s64 target_cycle, const double speed)
+{
+  // Calculate number of cycles since last throttle
+  const s64 cycles = target_cycle - m_throttle_last_cycle;
+  m_throttle_last_cycle = target_cycle;
+
+  // Based on number of cycles and emulation speed,
+  // increase the target deadline that we should throttle too
+  if (0.0 <= speed)
+    m_throttle_deadline += std::chrono::duration_cast<DT>((m_throttle_per_clock * cycles) / speed);
+
+  // Unless a maximum fallback is inplace, if the game runs at say 95% speed
+  // for 10s during a challenging section, as soon as a less challenging section
+  // comes up, it will run uncapped for a couple seconds. The max variance
+  // prevents this, while still allowing the throttler to maintain an average speed.
+  const DT max_fallback =
+      std::chrono::duration_cast<DT>(DT_ms(Config::Get(Config::MAIN_TIMING_VARIANCE)));
+
+  const TimePoint time = Clock::now();
+  const TimePoint min_deadline = time - max_fallback;
+  const TimePoint max_deadline = time + max_fallback;
+
+  if (m_throttle_deadline < min_deadline)
+  {
+    DEBUG_LOG_FMT(COMMON, "System can not to keep up with timings! [relaxing timings by {} us]",
+                  DT_us(min_deadline - m_throttle_deadline).count());
+    m_throttle_deadline = min_deadline;
+
+    // If system is unable to maintain timings, vsync will prevent us from catching up.
+    // This limits the CPU timings to the GPU's timings / causes vsyncs.
+    if (!m_throttle_disable_vsync)
+    {
+      WARN_LOG_FMT(COMMON, "System Lag Detected! VSync Capability Disabled!");
+      m_throttle_disable_vsync = true;
+    }
+  }
+  else if (m_throttle_deadline > max_deadline)
+    m_throttle_deadline = max_deadline;
+
+  // If the deadline has already passed there is no point in sleeping or recording
+  // performance data. The next time it sleeps, it will be able to assume that it
+  // did not sleep before and record accurate emulation performance data.
+  if (m_throttle_deadline < time)
+    return;
+
+  // If we are now behind the deadline, it means that the CPU is maintaining its
+  // timings on its own and we are safe to engage in VSync if we want.
+  if (m_throttle_disable_vsync)
+  {
+    WARN_LOG_FMT(COMMON, "System Lag Resolved! VSync Capability Reenabled!");
```

Comment:
Shouldn't thisbe an INFO_LOG?

Reply:
I'm not sure but I an definitely update it.